### ISSUE TITLE
`CredentialRequestOptions` make otherwise valid values invalid in an undesirable way

### ISSUE THREAD

**Issue by bzbarsky** [w3c/webauthn#750]
**23-Jul-2018 NOTE: see STATUS SUMMARY here:** https://github.com/w3c/webauthn/issues/750#issuecomment-407238236

----
https://w3c.github.io/webauthn/#credentialrequestoptions-extension has:

    partial dictionary CredentialRequestOptions {
        PublicKeyCredentialRequestOptions      publicKey;
    };

and [`PublicKeyCredentialRequestOptions`](https://w3c.github.io/webauthn/#dictdef-publickeycredentialcreationoptions) has:

    required BufferSource                challenge;

What this means in practice is that if this part of the spec is implemented and https://github.com/heycam/webidl/issues/76 is fixed then `{}` stops being a valid value for a `CredentialRequestOptions`.

Assuming that `CredentialsContainer.prototype.get` should be usable to request non-public-key credentials, and that https://github.com/heycam/webidl/issues/76 will be fixed at some point, this is not ideal.

**Comment by equalsJeffH** [comment]
we discussed this on [17-Jan-2018 webauthn call](https://www.w3.org/2018/01/17-webauthn-minutes.html). This requires some thought and reading-up on https://github.com/heycam/webidl/issues/76.

During the discussion, we noted that we _could_ nominally remove the "required" stipulation for `PublicKeyCredentialRequestOptions.challenge`, and note in spec prose that if one does not pass in a challenge when calling `navigator.credentials.{create(),get()}`, that the latter calls will not succeed (in some fashion...throw an error? return a null credential/assertion?). 

Though we noted that such an approach is not ideal and we need to research and discuss this issue further.  

Note also I believe @jcjones noted that the resolution to https://github.com/heycam/webidl/issues/76 is already implemented in Mozilla's Gecko.

**Comment by bzbarsky** [comment]
More precisely, "a resolution".  Whether that's what ends up happening in the webidl spec is unclear at this point.

**Comment by equalsJeffH** [comment]
this issue is essentially un-triaged (no assignee, no milestone) -- at least which milestone do we feel it ought to be assigned to?

**Comment by equalsJeffH** [comment]
Ok, @bzbarsky, I've been [RFTM](https://en.wiktionary.org/wiki/RTFM)'g here and have some questions...

Gecko implements this proposed resolution to https://github.com/heycam/webidl/issues/76:
> 1. A dictionary is "elidable" (better-naming-wanted) if it is either empty or only has members whose values are themselves elidable dictionaries.
> 2. When converting a dictionary to a JS value, skip over members whose values are elidable dictionaries, just like we skip over members that are not present.
> 
> For dictionaries that have default values that will still cause them to appear (with those default values)

..yes?

Assuming that what gecko implements (above?) is what is decided upon as the solution to heycam/webidl#76, **_AND_** if I understand the latter issue (and the related moz bugs) correctly, then I note that [`PublicKeyCredentialRequestOptions`](https://w3c.github.io/webauthn/#dictdef-publickeycredentialcreationoptions) et al has more `required` members than just `challenge`:
```
dictionary PublicKeyCredentialCreationOptions {
    required PublicKeyCredentialRpEntity         rp;
    required PublicKeyCredentialUserEntity       user;

    required BufferSource                             challenge;
    required sequence<PublicKeyCredentialParameters>  pubKeyCredParams;

    unsigned long                                timeout;
    sequence<PublicKeyCredentialDescriptor>      excludeCredentials = [];
    AuthenticatorSelectionCriteria               authenticatorSelection;
    AttestationConveyancePreference              attestation = "none";
    AuthenticationExtensionsClientInputs         extensions;
};

dictionary PublicKeyCredentialEntity {
    required DOMString    name;
    USVString             icon;
};

dictionary PublicKeyCredentialRpEntity : PublicKeyCredentialEntity {
    DOMString      id;
};

dictionary PublicKeyCredentialUserEntity : PublicKeyCredentialEntity {
    required BufferSource   id;
    required DOMString      displayName;
};

dictionary PublicKeyCredentialParameters {
    required PublicKeyCredentialType      type;
    required COSEAlgorithmIdentifier      alg;
};

```
..then it sounds like there are two more-or-less plausible approaches for the webauthn spec (again, [IIUC](https://en.wiktionary.org/wiki/IIUC)):

1. Remove the WebIDL "required" stipulation for all `required` `fooOptions` dictionary members (see above), and note in spec prose that if one does not pass all of them in with values when calling `navigator.credentials.{create(),get()}`, that the latter calls will not succeed (in some fashion: throw an error? return a null credential/assertion?). 

2.  Provide default values for all `required` `fooOptions` dictionary members, for example:
```
dictionary PublicKeyCredentialCreationOptions {
    required PublicKeyCredentialRpEntity         rp  = {};
    required PublicKeyCredentialUserEntity       user = {}; 
    required BufferSource                        challenge = null;
   [...]
};

dictionary PublicKeyCredentialEntity {
    required DOMString    name  =  "";
    USVString             icon;
};

dictionary PublicKeyCredentialUserEntity : PublicKeyCredentialEntity {
    required BufferSource   id = null;
    required DOMString      displayName = "";
};
```
..(I'm not sure if the above syntactically correct). This would cause those dictionaries with default values to "appear" (with those default values), e.g. if someone does..
```
CredentialRequestOptions     bar = {};
```
..yes?

I'm not sure whether this covers all the bases. Nor, if it does, which is the "better" option.

thoughts?

**Comment by bzbarsky** [comment]
> Gecko implements this proposed resolution to heycam/webidl#76:

Gecko doesn't implement the elidable stuff yet.

>  and note in spec prose that if one does not pass all of them in with values when calling navigator.credentials.{create(),get()}, that the latter calls will not succeed

Right, this is option 1.

> Provide default values for all required fooOptions dictionary members

Huh.  IDL allows a default value for required members in the grammar...  I filed https://github.com/heycam/webidl/issues/569 to fix that.

But even without that, the prose in https://heycam.github.io/webidl/#required-dictionary-member explicitly says:

    A required dictionary member must not have a default value.

because there are no sane semantics for a required thing with a default value...  That said, you could do non-required members with default values.  That would be option 2.

Option 3 is to argue that https://github.com/heycam/webidl/issues/76 should be wontfixed and that dictionary members of dictionaries should not be treated the same way as dictionaries as arguments.  Basically comes down to what a dictionary "is", conceptually...

**Comment by emlun** [comment]
Why should the default value for non-required members be the empty dictionary rather than `undefined` or `null`? It seems to me like if something is optional with no default value, then the default should be that it isn't there at all.

It seems to me like that behaviour would break all sorts of things. To take an example from the WebAuthn spec, [`CollectedClientData`][ccd] has this definition:

```
dictionary CollectedClientData {
    required DOMString           type;
    required DOMString           challenge;
    required DOMString           origin;
    TokenBinding                 tokenBinding;
};

dictionary TokenBinding {
    required TokenBindingStatus status;
    DOMString id;
};
```

Now, let's assume that heycam/webidl#76 is implemented and say I create a `CollectedClientData` instance and do not set the`tokenBinding` member. Since `CollectedClientData` is a dictionary and the `tokenBinding` member is not required, this would mean that the `tokenBinding` member in the result will be set to the empty dictionary. But since that dictionary is of type [`TokenBinding`][tb] which has a required member, the empty dictionary is an invalid value.

This is probably just another example of the very problem pointed out in the OP, but this example is much easier for me to think about. I imagine this kind of construction (an optional value with a required subcomponent if it's present) is _very_ common, and it seems to me like heycam/webidl#76 would invalidate _all_ such constructions. So unless I'm mistaken somewhere in my analysis (please point it out if I am), I think heycam/webidl#76 should be wontfixed.

[ccd]: https://w3c.github.io/webauthn/#dictdef-collectedclientdata
[tb]: https://w3c.github.io/webauthn/#dictdef-tokenbinding

**Comment by equalsJeffH** [comment]
@bzbarsky replied:
>> Gecko implements this proposed resolution to heycam/webidl#76:
> 
> Gecko doesn't implement the elidable stuff yet.

Hm, I was interpreting your https://github.com/w3c/webauthn/issues/750#issuecomment-358422136 in combination with https://github.com/heycam/webidl/issues/76#issue-118108356 and @jcjones's comment on the webauthn call, to mean that Gecko does implement the "elidable stuff".

You implied gecko implements "a resolution" and are now indicating that the aforementioned resolution is different than the "elidable stuff" ?

>> and note in spec prose that if one does not pass all of them in
>> with values when calling navigator.credentials.{create(),get()},
>> that the latter calls will not succeed
> 
> Right, this is option 1.

ok.

> ... the prose in
> https://heycam.github.io/webidl/#required-dictionary-member
> explicitly says:
> 
> A required dictionary member must not have a default value.

doh.

> you could do non-required members with default values.  That would be option 2.

ah!  so we can do this:
```
dictionary PublicKeyCredentialCreationOptions {
    PublicKeyCredentialRpEntity         rp  = {};
    PublicKeyCredentialUserEntity       user = {}; 
    BufferSource                        challenge = null;
    [...]
};

dictionary PublicKeyCredentialEntity {
    DOMString     name  =  "";
    USVString     icon;
};

dictionary PublicKeyCredentialUserEntity : PublicKeyCredentialEntity {
    BufferSource   id = null;
    DOMString      displayName = "";
};
```
..?   (I simply removed all the `required` keywords).


> Option 3 is to argue that https://github.com/heycam/webidl/issues/76
> should be wontfixed and that dictionary members of dictionaries
> should not be treated the same way as dictionaries as arguments.
> Basically comes down to what a dictionary "is", conceptually...

yes, this seems to be an interesting question to discuss over there.  @emlun followed up his above comment here with https://github.com/heycam/webidl/issues/76#issuecomment-401766639

**Comment by bzbarsky** [comment]
> I imagine this kind of construction (an optional value with a required subcomponent if it's present) is very common

So far we have two specs trying to do it: webauthn and some proposals for webrtc.  So no, it's not very common.  Maybe because `required` is somewhat new?

> You implied gecko implements "a resolution" and are now indicating that the aforementioned resolution is different than the "elidable stuff" ?

Gecko implements a specific behavior which is one of the proposals under consideration in <https://github.com/heycam/webidl/issues/76>: optional dictionaries always default to null and no other changes from the current WebIDL spec.  Adding the elidable thing is an additional proposal which would affect various dictionary-to-JS conversions, possibly including already-shipping ones.  I haven't checked to see whether there are currently dictionaries returned to JS that include nested dictionaries that would be affected.

> ah! so we can do this:

Sort of.  You can't use `{}` as a default value; see grammar at <https://heycam.github.io/webidl/#prod-DefaultValue>.  But you can use `null`, which will have the same effect.

**Comment by emlun** [comment]
>So far we have two specs trying to do it: webauthn and some proposals for webrtc.  So no, it's not very common.  Maybe because `required` is somewhat new?

Ah, ok. Though I wasn't thinking only about specs, but also API design in a broader sense. For example, I've worked in a project where we used [JSON schema][json-schema] to declare an API where these kinds of required members nested in optional members were quite frequent. It seems to me like a very natural thing to do with a language as expressive as JS literals are - and if people do it in APIs I think they'll expect it to work the same in specs. So if this is a new feature I imagine that this usage will only grow more common over time.

[json-schema]: http://json-schema.org/latest/json-schema-core.html

**Comment by bzbarsky** [comment]
> "optional dictionaries" in the dictionaries-as-optional-function-args case?

In all cases when the value is normally optional.  That means optional function args and dictionary members.

> Ok, so are you saying that we can address this issue by updating our spec in this fashion:

You can, yes.

I really think it's worth seeing exactly what happens in  https://github.com/heycam/webidl/issues/76 before making changes here, though.

**Comment by equalsJeffH** [comment]
Thx for clarifications @bzbarsky.   
> I really think it's worth seeing exactly what happens in heycam/webidl#76 before making changes here, though.
That sounds prudent. So we can punt on this issue for now ?  We are trying to attain Proposed Rec soon and Rec before calendar-year-end.

**Comment by bzbarsky** [comment]
Does the spec actually describe what should happen when the .publicKey dictionary is missing?  As far as I can tell https://w3c.github.io/webauthn/#createCredential asserts it's present in the CredentialCreationOptions, but it's not clear to me what guarantees that, exactly...  It's also not clear to me when the "publicKey" member of CredentialRequestOptions ever gets used.

As long as the spec completely describes behavior when the dictionaries involved are missing, it should be fine as things stand.

**Comment by bzbarsky** [comment]
(Oh, and as long as the behavior is reasonable; there are some issues with webappsec-credential-management doing stuff in parallel sections that might involve buffer sources given  the dictionaries this spec defines.)

**Comment by emlun** [comment]
As far as I understand, the presence of the `publicKey` member is what triggers the client to invoke the WebAuthn methods instead of other Credential Management backends.

**Comment by bzbarsky** [comment]
@emlun Where is that specified?  That's the part I was trying to find and failing.

**Comment by emlun** [comment]
Looks like it's defined in CredMan, loosely [here](https://w3c.github.io/webappsec-credential-management/#credentialcreationoptions-dictionary) and in more detail [here](https://w3c.github.io/webappsec-credential-management/#implementation-extension).

**Comment by bzbarsky** [comment]
Right, but where in the processing model is that determination of what to call actually made?

**Comment by emlun** [comment]
I _think_ it happens [here](https://w3c.github.io/webappsec-credential-management/#algorithm-create), where _interfaces_ refers to [here](https://w3c.github.io/webappsec-credential-management/#credentialrequestoptions-relevant-credential-interface-objects) where _options_ is the options argument object sent to the `create()` method and the `[[type]]` slot for the [PublicKeyCredential][pkc] interface contains `"public-key"`.

...actually, does that mean that the [`publicKey`][publicKey] member name in `CredentialCreationRequestOptions` should also be `"public-key"`, or vice-versa? @equalsJeffH @selfissued Who's our expert on this?

[pkc]: https://w3c.github.io/webauthn/#publickeycredential
[publicKey]: https://w3c.github.io/webauthn/#dom-credentialcreationoptions-publickey

**Comment by equalsJeffH** [comment]
> I _think_ it happens [here](https://w3c.github.io/webappsec-credential-management/#algorithm-create), where _interfaces_ refers to [here](https://w3c.github.io/webappsec-credential-management/#credentialrequestoptions-relevant-credential-interface-objects) where _options_ is the options argument object sent to the `create()` method and the `[[type]]` slot for the [PublicKeyCredential][pkc] interface contains `"public-key"`.

[IIUC](https://en.wiktionary.org/wiki/IIUC), the above is correct.

> ...actually, does that mean that the publicKey member name in CredentialCreationRequestOptions should also be "public-key", or vice-versa?

IIUC, I do not think that the difference between the `[[type]]` slot's value of `public-key` and the  [CredentialCreationOptions](https://www.w3.org/TR/webauthn/#credentialcreationoptions-extension) and [CredentialRequestOptions](https://www.w3.org/TR/webauthn/#credentialrequestoptions-extension) dictionare extensions' member name of `publicKey` causes any issues. 

cc: @mikewest

**Comment by bzbarsky** [comment]
@emlun Thank you!  That does look like the right thing now, and it's doing the "is missing" check I was kinda looking for.

But I think  you're right about the name mismatch being a problem.  @equalsJeffH how could it not  cause  issues?  The algorithm checks whether the dictionary has a "public-key" member, which it never does if the dictionary member is called "publicKey".

**Comment by equalsJeffH** [comment]
thanks @bzbarsky, as I said, [IIUC](https://en.wiktionary.org/wiki/IIUC), which means I _might not understand correctly_, which seems to be the case here (sigh, sorry)...

Looking more closely at [credman](https://w3c.github.io/webappsec-credential-management/)'s magic [**relevant credential interface objects sub-algorithm**](https://w3c.github.io/webappsec-credential-management/#credentialrequestoptions-relevant-credential-interface-objects), i'm now thinking you & @emlun are correct, due to steps 4.2 & 4.3 in said sub-alg:
```
4. For each object in interface objects:
    1. [ . . . ]
    2. Let key be object’s [[type]] slot’s value.   // "public-key"
    3. If options[key] exists, append object to 
        relevant interface objects.                 // options[public-key] throws (?)
```

So, perhaps browser implementors can weigh-in on whether it matters if we change..

1.  the [CredentialCreationOptions](https://www.w3.org/TR/webauthn/#credentialcreationoptions-extension) and [CredentialRequestOptions](https://www.w3.org/TR/webauthn/#credentialrequestoptions-extension) dictionare extensions' member name of `publicKey` to be `public-key`, or 
2. the [PublicKeyCredentialType](https://www.w3.org/TR/webauthn/#credentialType)'s enumeration value to be `"publicKey"`   (rather than the present `"public-key"`

..?

cc @kpaulh @agl @akshayku @jcjones @ttaubert 

[publicKey]: https://w3c.github.io/webauthn/#dom-credentialcreationoptions-publickey

**Comment by emlun** [comment]
For what it's worth, (1) would require changes to RP code as well in addition to clients, while (2) would not. On the other hand, (2) looks like more idiomatic JS since a bare `public-key` is not a valid object key literal in JS (the string literal `"public-key"` is a valid key, though).

**Comment by equalsJeffH** [comment]
hm. Plus there's the orange note in [WebIDL S2.8. Enumerations](https://heycam.github.io/webidl/#idl-enums) that sez in part:
>  It is strongly suggested that enumeration values be all lowercase, and that multiple words be separated using dashes or not be separated at all, unless there is a specific reason to use another value naming scheme. For example, an enumeration value that indicates an object should be created could be named "createobject" or "create-object".

**Comment by bzbarsky** [comment]
Right, enumerations are meant to be values, not property names...

**Comment by equalsJeffH** [comment]
Ok, thx, so how do we fix this such that we can have our cake and eat it too, i.e., have  [credman](https://w3c.github.io/webappsec-credential-management/)'s magic [**relevant credential interface objects sub-algorithm**](https://w3c.github.io/webappsec-credential-management/#credentialrequestoptions-relevant-credential-interface-objects) work for us?  Or, is this a credman flaw? 

Apparently, it's primarily a spec-level issue because there's running impls of this stuff...

**Comment by bzbarsky** [comment]
It's probably best to spin off the enumerations bit into a separate issue, since it's not really related to this issue.

And presumably impls are just doing something different from what the spec says and it's worth asking them what they do.

**Comment by equalsJeffH** [comment]
**STATUS SUMMARY**

**The above interstitial thread beginning with this comment** by @bzbarsky -- https://github.com/w3c/webauthn/issues/750#issuecomment-403167486, and continuing to the above comment https://github.com/w3c/webauthn/issues/750#issuecomment-404903787, _inclusive_ -- **is now forked over to issue #1004.**

This issue #750 has a proposed solution in this above comment: https://github.com/w3c/webauthn/issues/750#issuecomment-403125640, to which @bzbarsky suggests (in https://github.com/w3c/webauthn/issues/750#issuecomment-403137585) that we **wait until heycam/webidl#76 is resolved before applying any solution to this original issue**.

**Comment by bzbarsky** [comment]
I think you should consider https://github.com/heycam/webidl/issues/76 resolved (as "no change to the webidl spec") for purposes of this discussion.  The only question left open there does not affect this spec.

**Comment by equalsJeffH** [comment]
In discussion on webauthn call, it seems that we ought to add to the beginning of #createCredential and #getAssertion algs a step that checks for "required" options dictionary memebers' presence and bail out if any are not present.  Also need to make determination whehter we wish to convert present "required" dict members to having defalt NULL values, or no default values (but with comments indicating "requiredness").

**Comment by agl** [comment]
In Chromium, we have set the members of `CredentialCreationOptions` as nullable ([source](https://cs.chromium.org/chromium/src/third_party/blink/renderer/modules/credentialmanager/credential_creation_options.idl)). Thus if a given member isn't present, it's null, and then everything works out.

That contrasts with [what's in Webauthn](https://w3c.github.io/webauthn/#credentialcreationoptions-extension). Perhaps Webauthn should allow these fields to be nullable so that the required members within each subtype's options aren't a problem?

**Comment by bzbarsky** [comment]
> In Chromium, we have set the members of CredentialCreationOptions as nullable (source)

That's invalid IDL that only works due to  a bug in Chrome's IDL parser.  See  <https://bugs.chromium.org/p/chromium/issues/detail?id=890375>.

I don't understand what's wrong with my proposal from https://github.com/w3c/webauthn/issues/750#issuecomment-424946912 exactly.  Webauthn's IDL is fine at this point.   It just needs to clearly define what happens when those members are not present.  It's not clear to me that this is clearly specified right now.

**Comment by equalsJeffH** [comment]
@bzbarsky -- just to clarify:
> Webauthn's IDL is fine at this point. It just needs to clearly define what happens when those members are not present.

So are you saying that we can leave WebAuthn's IDL written as it presently is, with various dictionary members declared with the `required` keyword?  

And also we ought to add to the beginning of each of [#createCredential](https://w3c.github.io/webauthn/#createCredential) and [#getAssertion](https://w3c.github.io/webauthn/#discover-from-external-source) algs a step that checks for "required" options dictionary members' presence and bail out (with some appropriate error) if any are not present ?

**Comment by bzbarsky** [comment]
> So are you saying that we can leave WebAuthn's IDL written as it presently is, with various dictionary members declared with the required keyword?

Yes.

> And also we ought to add to the beginning of each of #createCredential and #getAssertion algs a step that checks for "required" options dictionary members' presence

Not quite, if I understand correctly how this works.  So to take a concrete example, let's look at `CredentialCreationOptions`.  The spec needs to define what happens when the `publicKey` member is not present.   If it _is_ present, then Web IDL will guarantee that its required members (e.g. `rp`) are also present.

I _think_  the idea is to not call the relevant  #createCredential at all if the corresponding member is not present in `CredentialCreationOptions`,  right?   As long as that's clearly defined, you should be all good.

This will give a behavior where passing `{}` or  `{ publicKey: undefined }` for the `CredentialCreationOptions` will not invoke the relevant #createCredential, while passing  `{ publicKey: null }` will throw.   I don't know whether that matches what Chrome does right now.

**Comment by emlun** [comment]
>The spec needs to define what happens when the `publicKey` member is not present. [...] I _think_ the idea is to not call the relevant #createCredential at all if the corresponding member is not present in `CredentialCreationOptions`

Yes, I think that behaviour is implied by the CredMan spec (see https://github.com/w3c/webauthn/issues/750#issuecomment-403434610 and the subsequent few comments), so I don't think we really need to do much more here. As far as I can tell, the one thing we _do_ need to figure out in order to integrate correctly with CredMan on this is to align the `"public-key"` type value and the `publicKey` member name.

**Comment by equalsJeffH** [comment]
@bzbarsky wrote:
> I think the idea is to not call the relevant #createCredential at all if the corresponding member is not present in CredentialCreationOptions, right?

Correct, as @emlun [notes above](https://github.com/w3c/webauthn/issues/750#issuecomment-426970949).

@emlun wrote:
>  I don't think we really need to do much more here.

Agreed, given @bzbarsky's [comment above](https://github.com/w3c/webauthn/issues/750#issuecomment-424946912).

>  As far as I can tell, the one thing we do need to figure out in order to integrate correctly with CredMan on this is to align the "public-key" type value and the publicKey member name.

Agreed. As I [noted above](https://github.com/w3c/webauthn/issues/750#issuecomment-407238236), that issue is now forked over to issue #1004.

So, are we good with closing this issue #750?

**Comment by bzbarsky** [comment]
I am, yes.

**Comment by equalsJeffH** [comment]
ok, closing per discussion on 14-nov-2018 call
